pkg/volume/util.go

```diff
@@ -345,3 +346,42 @@ func Zones2Set(zonesString string) (sets.String, error) {
 	}
 	return zonesSet, nil
 }
+
+// ValidatePVCSelector validates Selector part of a PVC:
+// - in case there is no Selector the PVC is valid
+// - makes sure that only allowedKeys are present in the Selector matchLabels part
+// - makes sure that only allowedKeys and allowedOperators are present in the Selector matchExpressions part
+// Return value:
+// - (true, nil) means PVC is valid (error == nil) and there is NO Selector OR (NO matchLabels AND NO matchExpressions) (bool == true)
+// - (false, nil) means PVC is valid (error == nil) and there is at least a value in matchLabels or matchExpressions specified (bool == false)
+// - (false, error) means PVC is not valid
+// - (true, error) shall never happen
+func ValidatePVCSelector(pvc *v1.PersistentVolumeClaim) (bool, error) {
+	allowedKeys := map[string]bool{unversioned.LabelZoneFailureDomain: true, unversioned.LabelZoneRegion: true}
+	allowedOperators := map[unversioned.LabelSelectorOperator]bool{unversioned.LabelSelectorOpIn: true, unversioned.LabelSelectorOpNotIn: true}
+	if pvc.Spec.Selector == nil {
+		return true, nil
+	}
+	if len(pvc.Spec.Selector.MatchExpressions) < 1 && len(pvc.Spec.Selector.MatchLabels) < 1 {
+		return true, nil
+	}
+	if len(pvc.Spec.Selector.MatchLabels) > 0 {
+		for label := range pvc.Spec.Selector.MatchLabels {
+			if !allowedKeys[label] {
+				return false, fmt.Errorf("Persistent volume claim (%v), matchLabels part: the %v key is not permitted", pvc.Name, label)
```

Comment:
- there is no need to introduce `Persistent volume claim (%v),` in the error messages, volume controller will add it and it's quite long already. **This applies to all errors reported in all patches in this PR**

- `matchLabels part: the %v key is not permitted` -> `key %q is not permitted in selector.matchLabels` to produce a sentence and emphasize what's the key here. Applies to all errors below in this patch.